bindings/rust-examples/client-hello-config-resolution/src/bin/server.rs

```diff
@@ -0,0 +1,114 @@
+use s2n_tls::{
+    callbacks::{ClientHelloCallback, ConnectionFuture},
+    security::{Policy, DEFAULT_TLS13},
+};
+use s2n_tls_tokio::TlsAcceptor;
+use std::{
+    collections::HashMap,
+    error::Error,
+    net::{IpAddr, Ipv4Addr},
+    pin::Pin,
+};
+use tokio::{io::AsyncWriteExt, net::*};
+
+const PORT: u16 = 1738;
+
+/// Used by the server to resolve the appropriate config
+pub struct AnimalConfigResolver {
+    // this stores the mapping from sni -> config
+    configs: HashMap<String, s2n_tls::config::Config>,
+}
+
+impl Default for AnimalConfigResolver {
+    fn default() -> Self {
+        let mut configs = HashMap::new();
+        configs.insert("www.wombat.com".to_owned(), server_config("wombat"));
+        configs.insert("www.kangaroo.com".to_owned(), server_config("kangaroo"));
+        Self { configs }    }
+}
+
+impl ClientHelloCallback for AnimalConfigResolver {
+    fn on_client_hello(
+        &self,
+        connection: &mut s2n_tls::connection::Connection,
+    ) -> Result<Option<Pin<Box<dyn ConnectionFuture>>>, s2n_tls::error::Error> {
+        let sni = match connection.server_name() {
+            Some(sni) => sni,
+            None => {
+                println!("connection contained no SNI");
+                return Err(s2n_tls::error::Error::application("no sni".into()));
+            }
+        };
+        let config_ref = match self.configs.get(sni) {
+            Some(c) => c,
+            None => {
+                println!("unknown sni: {sni}");
+                return Err(s2n_tls::error::Error::application("unknown sni".into()));
+            }
+        };
+        println!("setting connection config associated with {sni}");
+        let config = config_ref.clone();
+        connection.set_config(config).unwrap();
+        // Ok -> the function completed successfully
+        // None -> s2n-tls doesn't need to poll this to completion
+        Ok(None)
```

Comment:
I wonder if you should document that the [ConfigResolver](https://github.com/aws/s2n-tls/blob/main/bindings/rust/s2n-tls/src/callbacks/client_hello.rs#L39) can be used if a future is needed / this needs to be async

Reply:
Added a comment for now, will follow up this PR adding an example for it [personal branch](https://github.com/jmayclin/s2n-tls/blob/async-load-example/bindings/rust-examples/client-hello-config-resolution/src/bin/async_load_server.rs)